airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/registry_entry.py

```diff
@@ -555,14 +576,56 @@ def metadata_entry(context: OpExecutionContext) -> Output[Optional[LatestMetadat
     return Output(value=metadata_entry, metadata=dagster_metadata)
 
 
+def find_release_candidates_for_metadata_entry(
+    metadata_entry: LatestMetadataEntry, release_candidate_metadata_entries: Optional[List[LatestMetadataEntry]]
+) -> Optional[List[LatestMetadataEntry]]:
+    """Find the release candidates metadata definition for the given metadata entry.
+
+    Args:
+        metadata_entry (LatestMetadataEntry): The metadata entry.
+        release_candidate_metadata_entries (List[LatestMetadataEntry]): The release candidate metadata definitions.
+
+    Returns:
+        Optional[List[LatestMetadataEntry]]: The release candidates metadata definition.
+    """
+    if not release_candidate_metadata_entries:
+        return None
+
+    current_semver_version = semver.Version.parse(metadata_entry.metadata_definition.data.dockerImageTag)
+
+    # find the release candidate metadata definition that matches the given metadata entry
+    matching_release_candidates = [
+        release_candidate_metadata
+        for release_candidate_metadata in release_candidate_metadata_entries
+        if release_candidate_metadata.metadata_definition.data.dockerRepository == metadata_entry.metadata_definition.data.dockerRepository
+        # Only consider release candidates with a higher version than the metadata entry
+        and semver.Version.parse(release_candidate_metadata.metadata_definition.data.dockerImageTag) > current_semver_version
+    ]
+
+    if not matching_release_candidates:
+        return None
+
+    # We currently want to enforce upstream that there's always only one release candidate per connector
+    # But in case this is not the case, we'll take the one with the closest version to the current one
+    sorted_release_candidates = sorted(
+        matching_release_candidates, key=lambda x: semver.Version.parse(x.metadata_definition.data.dockerImageTag)
+    )
+    # We keep it a single item list until we have a use case for multiple release candidates
+    return [sorted_release_candidates[0]]
+
+
 @asset(
-    required_resource_keys={"slack", "root_metadata_directory_manager"},
+    required_resource_keys={"slack", "root_metadata_directory_manager", "release_candidate_metadata_file_blobs"},
     group_name=GROUP_NAME,
     partitions_def=metadata_partitions_def,
     auto_materialize_policy=AutoMaterializePolicy.eager(max_materializations_per_minute=MAX_METADATA_PARTITION_RUN_REQUEST),
 )
 @sentry.instrument_asset_op
-def registry_entry(context: OpExecutionContext, metadata_entry: Optional[LatestMetadataEntry]) -> Output[Optional[dict]]:
+def registry_entry(
+    context: OpExecutionContext,
+    metadata_entry: Optional[LatestMetadataEntry],
+    release_candidate_metadata_entries: Optional[List[LatestMetadataEntry]],
```

Comment:
Here:
- `metadata_entry` corresponds to a metadata entry partition
- `release_candidate_metadata_entries` to an asset with all RC metadata

Matching a metadata entry to its release candidate metadata file is done in `find_release_candidate_for_metadata_entry`.

@bnchrch I'm not 💯 this is the optimal approach if `release_candidate_metadata_entries` is growing.
Would there be a mechanism, leveraging partitions, which could make this `registry_entry` function take a `metadata_entry` and a `release_candidate_metadata_entry` partition? In other word the matching would happen elsewhere on partition generation.
In other words: I'm not sure what happens if an assets takes inputs from two different partition definitions

Reply:
Hmm yeah, so this is awkward because (I think) if 
1. A RC metadata gets removed or added
2. Then all registry entries are reprocessed

(Can you confirm Im right about that?)

If I am then we have three options to fix
1. We create multiple partition keys. So that the registry entry partitions key is a composite of `latest/version etag and RC etag`. This will be messy and complex. I dont recommend
2. On new RC metadata, after processing, we delete the partition key for its corresponding latest to force a reprocess. This also feels wrong.
3. We hoist the release candidate processing to the registry level, since we want that regenerated on each new RC anyway.

I think #3 makes the most sense

Reply:
Oh yes you are right. The `AutoMaterialization` leads to re-generation of all entries when the `release_candidate_metadata_entries` is re-materialized... This is definitely not what we want. 

> We hoist the release candidate processing to the registry level,

I'm not sure I get what you mean there. Aren't we already at the registry level as this assets generate registry entries. You might mean to do it when we aggregated per connector registry entry in a the global registry. Sounds indeed possible. Will do it.